libraries/AP_IOMCU/iofirmware/analog.cpp

```diff
@@ -84,5 +85,6 @@ uint16_t adc_sample_vservo(void)
 uint16_t adc_sample_vrssi(void)
 {
     const uint32_t channel = ADC_SQR3_SQ1_N(ADC_CHANNEL_IN5);
-    return adc_sample_channel(channel) * 9900 / 4096;
+    // scale to volts and scale to mV for transport
+    return adc_sample_channel(channel) * VOLTAGE_SCALING * 1000;
```

Comment:
This doesn't look better, `VOLTAGE_SCALING` is a float. You are taking what was a pure integer calculation before, casting it to float, then casting back to integer to return from the function. I don't think this has helped readability enough to justify 2 float conversions when it can only lower the accuracy of the returned result.

Reply:
As an added fun thing that I just realized, the method type is uint16. So that means the current code is already increasing the quantization error by rounding the mV measurement. That isn't a big deal because not many of our things rely on mV level precision. I also haven't dug deep enough nor done the math to determine the total uncertainty of the STM32 ADC.

What you are saying is that the IOMCU should be doing purely integer math. That would mean that the data sent to the FMU would be in ADC steps (uint16). Then, with the power of an FPU, do the conversion to Volts (float). That would preserve the resolution of the measure measurement and minimize the quantization error to where it should be.

Reply:
@WickedShell I think the math is correct now for shifting where the calculation are done. Please take a look if you get a chance.